man/vcffilter.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "VCFFILTER" "1" "" "vcffilter (vcflib)" "vcffilter (VCF filter)"
.hy
.SH NAME
.PP
\f[B]vcffilter\f[R]
.SH SYNOPSIS
.PP
\f[B]vcffilter\f[R] options
.SH DESCRIPTION
.PP
VCF filter the specified vcf file using the set of filters
.SH OPTIONS
.IP
.nf
\f[C]


options:
    -f, --info-filter     specifies a filter to apply to the info fields of records,
                          removes alleles which do not pass the filter
    -g, --genotype-filter specifies a filter to apply to the genotype fields of records
    -k, --keep-info       used in conjunction with \[aq]-g\[aq], keeps variant info, but removes genotype
    -s, --filter-sites    filter entire records, not just alleles
    -t, --tag-pass        tag vcf records as positively filtered with this tag, print all records
    -F, --tag-fail        tag vcf records as negatively filtered with this tag, print all records
    -A, --append-filter   append the existing filter tag, don\[aq]t just replace it
    -a, --allele-tag      apply -t on a per-allele basis.  adds or sets the corresponding INFO field tag
    -v, --invert          inverts the filter, e.g. grep -v
    -o, --or              use logical OR instead of AND to combine filters
    -r, --region          specify a region on which to target the filtering, requires a BGZF
                          compressed file which has been indexed with tabix.  any number of
                          regions may be specified.

Filter the specified vcf file using the set of filters.
Filters are specified in the form \[dq]<ID> <operator> <value>:
 -f \[dq]DP > 10\[dq]  # for info fields
 -g \[dq]GT = 1|1\[dq] # for genotype fields
 -f \[dq]CpG\[dq]  # for \[aq]flag\[aq] fields

Operators can be any of: =, !, <, >, |, &

Any number of filters may be specified.  They are combined via logical AND
unless --or is specified on the command line.  Obtain logical negation through
the use of parentheses, e.g. \[dq]! ( DP = 10 )\[dq]

For convenience, you can specify \[dq]QUAL\[dq] to refer to the quality of the site, even
though it does not appear in the INFO fields.

type: filter
\f[R]
.fi
.SH EXIT VALUES
.TP
\f[B]0\f[R]
Success
.TP
\f[B]not 0\f[R]
Failure
.SH SEE ALSO
.PP
\f[B]vcflib\f[R](1)
.SH OTHER
.SH LICENSE
.PP
Copyright 2011-2026 (C) Erik Garrison and vcflib contributors.
MIT licensed.
Copyright 2020-2026 (C) Pjotr Prins.
.SH AUTHORS
Erik Garrison and vcflib contributors.


RELEASE_NOTES.md
For contributions see
[contributors](https://github.com/vcflib/vcflib/graphs/contributors)
and
[commits](https://github.com/vcflib/vcflib/commits/master).

## ChangeLog ongoing

* Added guix-container-manifest.scm that allows building Singularity and Docker containers(!)

## ChangeLog v1.0.15 (20260319)

* Fix iterator usage in vcfkeepinfo and vcfremoveaberrantgenotypes (Jukka Matilainen)
* Fixes for vcfwave (Glenn Hickey)
* Added include files to build with vg
* Added pkgconfig support
* Fix wfa2lib builds and sent upstream
* Update zig build to 0.15, or later, and fix static build with zig for vcfcreatemulti
* Create tar ball with static binaries
* Fix building with git submodules only
* Download singularity and docker containers from https://files.genenetwork.org/software/vcflib/v1.0.15/
  md5sum:
  5768540e7bef2db1363e422bc37614c4  vcflib-1.0.15-docker.tar.gz # Docker
  680cd7675325561636e03274e679d195  vcflib-1.0.15-singularity.pack.gz # Singularity
  7f60bb29dfb498c1059a33e10e99ac02  vcflib-static-git-v1.0.15-2-g0d40ff1.tar.gz # Static binaries

## ChangeLog v1.0.14 (20250518)

* CMake: Fixed building static libs and shared libs (the previous version just compiled a static lib)
* Updated zig compile for 0.14 (MichaelK)
* Lots of refactoring (MichaelK)
* CMake: Improved linker speeds by moving some libs
* CMake: Improved markdown/pydoc tests, so we don't show the path to the binary
* guix.scm has multiple outputs including static binaries and a test mode
* guix.scm has procedure for building with local htslib and tabixpp only
* Added @RedPenguin100 Michael Kovaliov as a maintainer (welcome!)

## ChangeLog v1.0.13 (20250417)

* Zig support was added to Guix and Conda - and we expect it to be more common soon
* Addind release protocol to README - so I don't forget bumping version etc.
* Fixed vcfwave issues (see https://github.com/vcflib/vcflib/issues/416) and conda build (thanks Michael Kovaliov, aka @RedPenguin100)
* Fixed github actions (Michael)
* Lots of minor refactoring and cleanup, as well as a memory leak (Michael)
* tabix.hpp fixed and set `WFA_INCLUDE_DIRS` and `-DBUILD_SHARED_LIBS=ON` (thanks Michael Cho, aka @cho-m)
* Cmake fixes (thanks Adam Novak)
* Updated simde submodule
* Fix CMakeLists.txt not to try to install wfa-lib when it is a package
* Reverted on Canocalize extraction in Variant class because of downstream use, see https://github.com/vcflib/vcflib/issues/417

## ChangeLog v1.0.12 (20241122)

* Improved parsing of INFO and FORMAT lines by @jeizenga in https://github.com/vcflib/vcflib/pull/374
* Support external tabixpp by @adamnovak in https://github.com/vcflib/vcflib/pull/375
* Fix vcfwave.cpp by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/407
* Use regex to find the '..' between postions and replace it with '-' by @debbyku in https://github.com/vcflib/vcflib/pull/405
* Fix tag-fail long option that was overriden by tag-pass in vcffilter.cpp by @Gullumluvl in https://github.com/vcflib/vcflib/pull/404
* `vcfwave`: fix condition to avoid nullifying valid SNPs and MNPs by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/408
* Upgraded Zig support to 0.13.0 by @pjotrp
* Merged multichoose code into vcflib since no one else uses it
* Added sources for canonicalize too
* Moved Fasta.h (fastahack) dependencies from Variant.h into sources
* Updated multichoose and simde modules
* Improved vcfwave support

### New Contributors
* @jeizenga made their first contribution in https://github.com/vcflib/vcflib/pull/374
* @debbyku made their first contribution in https://github.com/vcflib/vcflib/pull/405
* @Gullumluvl made their first contribution in https://github.com/vcflib/vcflib/pull/404

## ChangeLog v1.0.11-pre ()

+ Stopped vendoring wfa2lib by default - so now the Debian build command is
  `cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DZIG=OFF ..`


## ChangeLog v1.0.10 (20240420)

Vcflib maintenance release

+ Fixed vcfwave bugs - thanks Andrea Guarracino!
+ Fixed c++17 compatability - thanks Alex Petty!
+ Fixed tabixpp script - thanks Blaise Li!
+ Improved for vcfwave the wfa2lib build, fixed running tests with upstream build update
+ Removed deprecated binary_function
+ Removed unused files from repo
+ Updated README

## ChangeLog v1.0.9 (20230211)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.8 (20230210)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.7 (20230207)

Vcflib maintenance release - mostly for including in Debian

+ Fixed regression discovered by garguantua_kerr and atille of Debian (thanks!)
+ Added note on bio-vcf in vcffilter doc
+ notes on vcfcreatemulti and backtracking
+ CMake: honour include(GNUInstallDirs) paths (I forgot)

## ChangeLog v1.0.6 (20230129)

Vcflib maintenance release - mostly for including in Debian

+ Fixed zig complaining about leaking memory
+ Added CMake Debian support with -DWFA_GITMODULE=OFF
+ Introduced CMake include(GNUInstallDirs)
+ Successfully built wfa2 using embedded CMakeLists.txt
+ Cleaned up CMakeLists.txt removing comments etc.
+ Reintroduced vcfcreatemulti in legacy mode when ZIG=OFF (for Debian)

## ChangeLog v1.0.5 (20230116)

Vcflib's first *Humpty Dumpty* release: [vcfcreatemulti](./doc/vcfcreatemulti.md) is the natural companion to [vcfwave](./doc/vcfwave.md).

Often variant callers are not perfect.
**vcfwave** with its companion tool **vcfcreatemulti** can take an existing VCF file that contains multiple complex overlapping and even nested alleles and, unlike Humpty Dumpty, take them apart and put them together again.
Thereby, hopefully, creating sane VCF output that is useful for analysis and getting rid of false positives.

We created these tools by including the state-of-the-art [biWFA](https://github.com/smarco/WFA2-lib) wavefront aligner.

src/vcffilter.cpp
/*
    vcflib C++ library for parsing and manipulating VCF files

    Copyright © 2010-2020 Erik Garrison
    Copyright © 2020      Pjotr Prins

    This software is published under the MIT License. See the LICENSE file.
*/

#include "Variant.h"
#include "split.h"
#include <getopt.h>

using namespace std;
using namespace vcflib;

void printSummary(char** argv) {
  cerr << "vcflib filter the specified vcf file using the set of filters" << endl << endl
       << "usage: " << argv[0] << " [options] <vcf file>" << endl
         << endl
         << "options:" << endl
         << "    -f, --info-filter     specifies a filter to apply to the info fields of records," << endl
         << "                          removes alleles which do not pass the filter" << endl
         << "    -g, --genotype-filter specifies a filter to apply to the genotype fields of records" << endl
         << "    -k, --keep-info       used in conjunction with '-g', keeps variant info, but removes genotype" << endl
         << "    -s, --filter-sites    filter entire records, not just alleles" << endl
         << "    -t, --tag-pass        tag vcf records as positively filtered with this tag, print all records" << endl
         << "    -F, --tag-fail        tag vcf records as negatively filtered with this tag, print all records" << endl
         << "    -A, --append-filter   append the existing filter tag, don't just replace it" << endl
         << "    -a, --allele-tag      apply -t on a per-allele basis.  adds or sets the corresponding INFO field tag" << endl
         << "    -v, --invert          inverts the filter, e.g. grep -v" << endl
         << "    -o, --or              use logical OR instead of AND to combine filters" << endl
         << "    -r, --region          specify a region on which to target the filtering, requires a BGZF" << endl
         << "                          compressed file which has been indexed with tabix.  any number of" << endl
         << "                          regions may be specified." << endl
         << endl
         << "Filter the specified vcf file using the set of filters." << endl
         << "Filters are specified in the form \"<ID> <operator> <value>:" << endl
         << " -f \"DP > 10\"  # for info fields" << endl
         << " -g \"GT = 1|1\" # for genotype fields" << endl
         << " -f \"CpG\"  # for 'flag' fields" << endl
         << endl
         << "Operators can be any of: =, !, <, >, |, &" << endl
         << endl
         << "Any number of filters may be specified.  They are combined via logical AND" << endl
         << "unless --or is specified on the command line.  Obtain logical negation through" << endl
         << "the use of parentheses, e.g. \"! ( DP = 10 )\"" << endl
         << endl
         << "For convenience, you can specify \"QUAL\" to refer to the quality of the site, even" << endl
         << "though it does not appear in the INFO fields." << endl
       << endl << "type: filter" << endl;
    exit(0);
}

bool passesFilters(Variant& var,
		   vector<VariantFilter>& filters,
		   bool logicalOr, string alt = "") {
    for (vector<VariantFilter>::iterator f = filters.begin();
	 f != filters.end(); ++f) {
      string s = "";
      if (logicalOr) {
	if (alt.empty()) {
	  if (f->passes(var, s)) return true;
	} else {
	  if (f->passes(var, s, alt)) return true;
	}
      } else {
	if (alt.empty()) {
	  if (!f->passes(var, s)) return false;
	} else {
	  if (!f->passes(var, s, alt)) return false;
	}
      }
    }
    if (logicalOr)
      return false;
    else
      return true;
}


int main(int argc, char** argv) {

    int c;
    bool invert = false;
    bool logicalOr = false;
    bool filterSites = false;
    bool keepInfo = false;
    vector<string> infofilterStrs;
    vector<VariantFilter> infofilters;
    vector<string> genofilterStrs;
    vector<VariantFilter> genofilters;
    string tagPass = "";
    string tagFail = "";
    string filterSpec;
    string alleleTag;
    vector<string> regions;
    bool replaceFilter = true;

    if (argc == 1)
        printSummary(argv);

    while (true) {
        static struct option long_options[] =
            {
                /* These options set a flag. */
                //{"verbose", no_argument,       &verbose_flag, 1},
                {"help", no_argument, 0, 'h'},
                {"filter-sites", no_argument, 0, 's'},
                {"info-filter",  required_argument, 0, 'f'},
                {"genotype-filter",  required_argument, 0, 'g'},
                {"tag-pass", required_argument, 0, 't'},
                {"tag-fail", required_argument, 0, 'F'},
                {"append-filter", no_argument, 0, 'A'},
                {"allele-tag", required_argument, 0, 'a'},
                {"invert", no_argument, 0, 'v'},
                {"or", no_argument, 0, 'o'},
                {"region", required_argument, 0, 'r'},
                {"keep-info", no_argument, 0, 'k'},
                //{"length",  no_argument, &printLength, true},
                {0, 0, 0, 0}
            };
        /* getopt_long stores the option index here. */
        int option_index = 0;

        c = getopt_long (argc, argv, "hvAsof:g:kt:F:r:a:",
                         long_options, &option_index);

        /* Detect the end of the options. */
        if (c == -1)
            break;

        switch (c)
        {
        case 0:
            /* If this option set a flag, do nothing else now. */
            if (long_options[option_index].flag != 0)
                break;
            printf ("option %s", long_options[option_index].name);
            if (optarg)
                printf (" with arg %s", optarg);
            printf ("\n");
            break;

        case 'f':
            filterSpec += " " + string(optarg);
            infofilterStrs.push_back(string(optarg));
            break;

        case 's':
            filterSites = true;
            break;

        case 'a':
            alleleTag = optarg;
            break;

        case 'g':
            filterSpec += " genotypes filtered with: " + string(optarg);
            genofilterStrs.push_back(string(optarg));
            break;

        case 't':
            tagPass = optarg;
            break;

        case 'F':
            tagFail = optarg;
            break;

        case 'A':
            replaceFilter = false;
            break;

        case 'h':
            printSummary(argv);
            exit(0);
            break;

        case 'v':
            invert = true;
            break;

        case 'o':
            logicalOr = true;
            break;

        case 'r':
            regions.push_back(optarg);
            break;

        case 'k':
        	keepInfo = true;
        	break;

        case '?':
            /* getopt_long already printed an error message. */
            printSummary(argv);
            exit(1);
            break;

        default:
            abort ();
        }
    }

    if(filterSpec.empty()){
      cout << "Argument -g or -f is required" << endl;
      exit(1);
    }

    if (genofilterStrs.size() == 0 && keepInfo) {
      cout << "argument '-k' (--keep-info) requires a Genotype filter: ('-g')" << endl
	   << "i.e.: -g \"GT = 1|1\" -k" << endl;
      exit(1);
    }

    filterSpec = filterSpec.substr(1); // strip leading " "

    VariantCallFile variantFile;
    string inputFilename;
    if (optind == argc - 1) {
      inputFilename = argv[optind];
      variantFile.open(inputFilename);
    } else {
      variantFile.open(std::cin);
    }

    if (!variantFile.is_open()) {
        return 1;
    }

    for (vector<string>::iterator f = infofilterStrs.begin();
	 f != infofilterStrs.end(); ++f) {
      infofilters.push_back(VariantFilter(*f, VariantFilter::RECORD,
					  variantFile.infoTypes));
    }

    for (vector<string>::iterator f = genofilterStrs.begin();
	 f != genofilterStrs.end(); ++f) {
        genofilters.push_back(VariantFilter(*f, VariantFilter::SAMPLE,
					    variantFile.formatTypes));
    }

    vector<string> headerlines = split(variantFile.header, "\n");
    variantFile.header.clear();
    for (vector<string>::iterator l = headerlines.begin();
	 l != headerlines.end(); ++l) {
      if (!filterSpec.empty()
	  && (l->find("INFO")
	      != string::npos || l + 1 == headerlines.end())) {
	variantFile.header += "##filter=\"" + filterSpec + "\"\n";
	filterSpec.clear();
      }
      variantFile.header += *l + ((l + 1 == headerlines.end()) ? "" : "\n");
    }

    if (!tagPass.empty()) {
      variantFile.addHeaderLine("##FILTER=<ID="+ tagPass +",Description=\"Record passes the filters: " + filterSpec + ".\">");
    }

    if (!tagFail.empty()) {
      variantFile.addHeaderLine("##FILTER=<ID="+ tagFail +",Description=\"Record fails the filters: " + filterSpec + ".\">");
    }

    if (!alleleTag.empty()) {
        if (tagFail.empty()) {
            tagFail = "";
        }
        if (tagPass.empty()) {
	  tagPass = "PASS";
        }
        variantFile.addHeaderLine("##INFO=<ID="+ alleleTag +",Number=A,Type=String,Description=\"" + tagPass + " if this allele passes the filters, " + tagFail + " if not, filters are: " + filterSpec + ".\">");
    }

    cout << variantFile.header << endl;

    /*
      if (genofilters.empty()) {
      variantFile.parseSamples = false;
      }
    */

    if (filterSites) {
      variantFile.parseSamples = false;
    }

    Variant var(variantFile);

    vector<string>::iterator regionItr = regions.begin();

    do {
      if (!inputFilename.empty() && !regions.empty()) {
	string regionStr = *regionItr++;
	variantFile.setRegion(regionStr);
      }

      while (variantFile.getNextVariant(var)) {

	if (!genofilters.empty()) {
	  for (vector<VariantFilter>::iterator f
		 = genofilters.begin(); f != genofilters.end(); ++f) {
	    f->removeFilteredGenotypes(var, keepInfo);
	  }
	}
	if (!infofilters.empty()) {
	  if (filterSites) {
	    bool passes = passesFilters(var, infofilters, logicalOr);
	    if (invert) {
	      passes = !passes;
	    }
	    if (passes) {
	      if (!tagPass.empty()) {
		if (alleleTag.empty()) {
		  if (replaceFilter) {
		    var.filter.clear();
		    var.addFilter(tagPass);
		  } else {
		    var.addFilter(tagPass);
		  }
		} else {
		  var.info[alleleTag].clear();
		  for (vector<string>::iterator
			 a = var.alt.begin(); a != var.alt.end(); ++a) {
		    var.info[alleleTag].push_back(tagPass);
		  }
		}
	      } else {
		if (!var.originalLine.empty()) {
		  cout << var.originalLine << endl;
		} else {
		  cout << var << endl;
		}
	      }
	    } else {
	      if (!tagFail.empty()) {
		if (alleleTag.empty()) {
		  if (replaceFilter) {
		    var.filter.clear();
		    var.addFilter(tagFail);
		  } else {
		    var.addFilter(tagFail);
		  }
		} else {
		  var.info[alleleTag].clear();
		  for (vector<string>::iterator
			 a = var.alt.begin(); a != var.alt.end(); ++a) {
		    var.info[alleleTag].push_back(tagFail);
		  }
		}
	      }
	    }
	    if (passes && !tagPass.empty()) {
	      cout << var << endl;
	    } else if (!tagFail.empty()) {
	      cout << var << endl;
	    }
	  } else { // filter out alleles which pass
	    // removes the failing alleles
	    vector<string> failingAlts;
	    vector<string> passingAlts;
	    vector<bool> passes;
	    for (vector<string>::iterator a = var.alt.begin();
		 a != var.alt.end(); ++a) {
	      if (invert xor passesFilters(var, infofilters, logicalOr, *a)) {
		passingAlts.push_back(*a);
		passes.push_back(true);
	      } else {
		failingAlts.push_back(*a);
		passes.push_back(false);
	      }
	    }
	    if (tagPass.empty()) { // if there is no specified tag, just remove the failing alts
	      if (failingAlts.size() < var.alt.size()) {
		for (vector<string>::iterator a = failingAlts.begin(); a != failingAlts.end(); ++a) {
		  var.removeAlt(*a);
		}
		cout << var << endl;
	      }
	    } else { // otherwise, apply the tag
	      if (alleleTag.empty()) {
		if (!passingAlts.empty()) {
		  if (replaceFilter) {
		    var.filter.clear();
		    var.addFilter(tagPass);
		  } else {
		    var.addFilter(tagPass);
		  }
		} else {
		  if (replaceFilter) {
		    var.filter.clear();
		    if (!tagFail.empty()) {
		      var.addFilter(tagFail);
		    }
		  } else {
		    if (!tagFail.empty()) {
		      var.addFilter(tagFail);
		    }
		  }
		}
	      } else {
		var.info[alleleTag].clear();
		for (vector<bool>::iterator p = passes.begin();
		     p != passes.end(); ++p) {
		  if (*p) {
		    var.info[alleleTag].push_back(tagPass);
		  } else {
		    var.info[alleleTag].push_back(tagFail);
		  }
		}
	      }
	      // TODO
	      // here, if we don't use genotype filters, we shouldn't re-print the samples
	      // we haven't done anything to this part of the input.
	      cout << var << endl;
	    }
	  }
	} else {
	  if (genofilters.empty()) {
	    cout << variantFile.line << endl;
	  } else {
	    cout << var << endl;
	  }
	}
      }

    } while (regionItr != regions.end());

    return 0;

}
